[Merged by Bors] - feat: order isomorphisms act on flags

YaelDillies opened the pull request:
This is crucial to the study of abstract polytopes. For example, a regular abstract polytope is one whose automorphism group is transitive on its flags.

From MiscYD

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

b-mehta commented on Mathlib/Algebra/Order/Group/Action/Flag.lean:
```diff
@@ -0,0 +1,30 @@
+
+open scoped Pointwise
+
+variable {𝕆 α : Type*}
```
```suggestion
variable {α : Type*}
```

b-mehta reviewed (commented):
What's your application of this? You mention MiscYD but you don't seem to have provided a link, and github tells me that flags aren't mentioned in that project.

YaelDillies commented:
> What's your application of this? You mention MiscYD but you don't seem to have provided a link, and github tells me that flags aren't mentioned in that project. In particular, it's not clear to me from your PR that this action is something necessarily useful in mathlib, or that even shows up informally.

I removed the code from MiscYD in the latest commit https://github.com/YaelDillies/MiscYD/commit/85f2cc585c6139355a9ca9a4df9a6be0362c94e7 because it is not used in MiscYD anymore. The story of this code is that @vihdzp wrote some code on (abstract) polytopes in Lean 3 ages ago, I cleaned it up, then ported the best parts to Lean 4. This is one of the best parts.

b-mehta commented:
> > What's your application of this? You mention MiscYD but you don't seem to have provided a link, and github tells me that flags aren't mentioned in that project. In particular, it's not clear to me from your PR that this action is something necessarily useful in mathlib, or that even shows up informally.
> 
> I removed the code from MiscYD in the latest commit [YaelDillies/MiscYD@85f2cc5](https://github.com/YaelDillies/MiscYD/commit/85f2cc585c6139355a9ca9a4df9a6be0362c94e7) because it is not used in MiscYD anymore. The story of this code is that @vihdzp wrote some code on (abstract) polytopes in Lean 3 ages ago, I cleaned it up, then ported the best parts to Lean 4. This is one of the best parts.

Does this mean that there is no application for this code? From your message, I'm still not seeing an actual application, or an argument for why it's useful in mathlib.

YaelDillies commented:
As said above, there is an application, written in Lean 3: https://github.com/leanprover-community/mathlib3/compare/master...polytopes

I don't know the odds of the rest of the project ever being done, but it did take me a while to abstract this out of the original code, so I thought we might as well have it. [Last time I did so, it was used instantly](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/MulAction.20for.20OrderIso).

vihdzp commented:
I don't currently have plans to ressurect that (I've got a lot on my plate as of now), though I could be convinced if @ocornoc was interested in getting back to this.

b-mehta commented:
> As said above, there is an application, written in Lean 3: [leanprover-community/mathlib3@master...polytopes](https://github.com/leanprover-community/mathlib3/compare/master...polytopes)

Can you point me to where you mentioned this application earlier? In the future, as has been mentioned to you in the past, if you claim that the code is from a particular project, please include a direct reference. Furthermore, it is generally not a great idea to claim the code is from a project when it is explicitly not in that project. 

> it did take me a while to abstract this out of the original code, so I thought we might as well have it. 

I'm afraid that "it took me a while to write" isn't exactly great criterion for inclusion in mathlib. In the other case you reference, there's a clear reason why the action is useful, in this one I'm struggling to see it. 


Please add to the PR description some justification for why this is a useful instance for mathlib.

vihdzp commented:
The automorphism groups of abstract polytopes are the basis for basically the entire theory. A regular abstract polytope for instance is one whose automorphism group is transitive on its flags.

(Maybe this isn't great justification, since I already mentioned the abstract polytopes project is currently paused).

b-mehta commented:
Thanks for the justification here, Violeta!

> The story of this code is that @vihdzp wrote some code on (abstract) polytopes in Lean 3 ages ago, I cleaned it up, then ported the best parts to Lean 4. This is one of the best parts.

This sounds like co-authorship is warranted?

YaelDillies commented:
Violeta's code was an inspiration, but I wrote all the content in this PR from scratch 😄

b-mehta commented:
bors merge

b-mehta commented:
bors r-

b-mehta commented on Mathlib/Algebra/Order/Group/Action/Flag.lean:
```diff
@@ -0,0 +1,30 @@
+-/
+import Mathlib.Algebra.Group.Action.Pointwise.Set.Basic
+import Mathlib.Algebra.Order.Group.Action.End
+import Mathlib.Order.Chain
```
```suggestion
import Mathlib.Order.Preorder.Chain
```

b-mehta commented:
bah, looks like you'll need to merge master
